executor/src/witgen/machines/block_machine.rs

```diff
@@ -276,42 +241,31 @@ impl<T: FieldElement> BlockMachine<T> {
         {
             // All values on the left hand side are known, check if this is a query
             // to the last row.
-            let row_before = self.row;
-            self.row = self.rows() - 1;
-            let result = self.process_outer_query(fixed_data, left, right);
-
-            match result {
-                Ok(result) => {
-                    if result.is_complete() {
-                        return Ok(result);
-                    } else {
-                        // If the query is not complete, we roll back the row change and continue.
-                        // This might happen if the block machine just validates some input (rather
-                        // than computing an output) and hasn't been called on that input before.
-                        self.row = row_before;
-                    }
-                }
-                Err(e) => {
-                    // Non-recoverable error
-                    return Err(e);
-                }
+            let row = self.rows() - 1;
+
+            let current = &self.data[row as usize];
+            // We don't have the next row, because it would be the first row of the next block.
+            // We'll use a fresh row instead.
+            let next = self.row_factory.fresh_row();
+            let row_pair = RowPair::new(current, &next, row, fixed_data, false);
+
+            let result = identity_processor.process_backlink(left, right, &row_pair)?;
+
+            if result.is_complete() {
+                return Ok(result);
             }
         }
 
-        let outer_polys = left
-            .iter()
-            .flat_map(|r| {
-                r.as_ref()
-                    .ok()
-                    .map(|e| e.nonzero_coefficients().map(|(i, _)| i))
-            })
-            .flatten()
-            .collect::<BTreeSet<_>>();
-
         let old_len = self.rows();
         self.append_new_block(fixed_data.degree)?;
         let mut outer_assignments = EvalValue::complete(vec![]);
 
+        // While processing the block, we'll add an additional row which will be
+        // removed once the block is done. This is done to prevent infinite loops:
+        // If we would ignore updates to the last row, they would be computed over
+        // and over again.
```

Comment:
```suggestion
        // If we ignored updates to the last row, they would be computed over
        // and over again.
```

🤓

Reply:
Is this extra row added on the caller side?

Reply:
I'm trying to reconstruct how it was before:
- [When processing identities, values of this row where always unknown](https://github.com/powdr-labs/powdr/blob/0ba95b8774e9f18ba343f3897a7f1b349c42d473/executor/src/witgen/machines/block_machine.rs#L654-L659)
- [Updates to this row where always ignored](https://github.com/powdr-labs/powdr/blob/0ba95b8774e9f18ba343f3897a7f1b349c42d473/executor/src/witgen/machines/block_machine.rs#L443)

So, I think this must have led to infinite loops in some cases (because the same updates are computed over and over again), which was fine because we didn't fail hard then. Think it's cleaner now.

Reply:
Okay, probably fine for now. Would a cleaner long term solution be for blocks to have one offset per column? Then they would more naturally support not being rectangular?